File: Sources/AutoFlash/RegisteredFirmware.swift
import Foundation

// 事前に登録したローカルの.uf2ファイル。GitHubを経由せず直接書き込む。
struct RegisteredFirmware: Codable, Identifiable, Hashable {
    var id = UUID()
    var name = "New Firmware"
    var filePath = ""

    var url: URL? {
        filePath.isEmpty ? nil : URL(fileURLWithPath: filePath)
    }

    var fileName: String {
        url?.lastPathComponent ?? ""
    }

    var suggestedName: String? {
        guard let url else { return nil }
        let value = url.deletingPathExtension().lastPathComponent
        return value.isEmpty ? nil : value
    }

    var fileExists: Bool {
        guard let url else { return false }
        return FileManager.default.fileExists(atPath: url.path)
    }
}

enum RegisteredFirmwareSettings {
    private static let key = "registered.firmwares"

    static var firmwares: [RegisteredFirmware] {
        get {
            guard let data = UserDefaults.standard.data(forKey: key) else { return [] }
            return (try? JSONDecoder().decode([RegisteredFirmware].self, from: data)) ?? []
        }
        set {
            if let data = try? JSONEncoder().encode(newValue) {
                UserDefaults.standard.set(data, forKey: key)
            }
        }
    }
}

File: windows/AutoFlash/Models/RegisteredFirmware.cs
using System.IO;

namespace AutoFlash.Models;

// 事前に登録したローカルの.uf2ファイル。GitHubを経由せず直接書き込む。
public sealed class RegisteredFirmware
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "New Firmware";
    public string FilePath { get; set; } = "";

    public string FileName => FilePath.Length == 0 ? "" : Path.GetFileName(FilePath);

    public string? SuggestedName
    {
        get
        {
            if (FilePath.Length == 0) return null;
            var value = Path.GetFileNameWithoutExtension(FilePath);
            return value.Length == 0 ? null : value;
        }
    }

    public bool FileExists => FilePath.Length > 0 && File.Exists(FilePath);
}
